src/compiler/scala/tools/nsc/transform/async/AnfTransform.scala

```diff
@@ -136,19 +136,15 @@ private[async] trait AnfTransform extends TransformUtils {
           }
 
         case ValDef(mods, name, tpt, rhs) => atOwner(tree.symbol) {
-          // Capture current cursor of a non-empty `stats` buffer so we can efficiently restrict the
+          // Capture size of `stats` buffer so we can efficiently restrict the
           // `changeOwner` to the newly added items...
-          var statsIterator = if (currentStats.isEmpty) null else currentStats.iterator
+          val oldItemsCount = currentStats.length
 
           val expr = atOwner(currentOwner.owner)(transform(rhs))
 
-          // But, ListBuffer.empty.iterator doesn't reflect later mutation. Luckily we can just start
-          // from the beginning of the buffer
-          if (statsIterator == null) statsIterator = currentStats.iterator
-
           // Definitions within stats lifted out of the `ValDef` rhs should no longer be owned by the
           // the ValDef.
-          statsIterator.foreach(_.changeOwner((currentOwner, currentOwner.owner)))
+          currentStats.iterator.drop(oldItemsCount).foreach(_.changeOwner((currentOwner, currentOwner.owner)))
```

Comment:
OK, now I understand the previous comment. Iterators concat lazily, so this works also for initial empty:
```
scala> import collection.mutable.ListBuffer
import collection.mutable.ListBuffer

scala> val xs = ListBuffer.empty[Int]
val xs: scala.collection.mutable.ListBuffer[Int] = ListBuffer()

scala> xs += 42
val res0: xs.type = ListBuffer(42)

scala> val n = xs.length
val n: Int = 1

scala> val ys = Iterator.empty[Int] ++ xs.iterator.drop(n)
val ys: Iterator[Int] = <iterator>

scala> xs += 27
val res1: xs.type = ListBuffer(42, 27)

scala> ys.next()
val res2: Int = 27
```
The `current.length` is captured by the iterator, not in a variable.

Reply:
that would also work, yes